tokio-trace-core/src/dispatcher.rs

```diff
@@ -66,6 +72,9 @@ impl Dispatch {
     /// [`Subscriber`]: ::Subscriber
     /// [`Event`]: ::Event
     pub fn as_default<T>(&self, f: impl FnOnce() -> T) -> T {
+        if thread::panicking() {
```

Comment:
Same, this can probably be done using `try_with` instead of checking for panicking.